Adding keymanagement for smimekeys additionally for identities

boekhorstb1 opened the pull request:
Ok I found some solution for adding multiple smimekeys, also for identities. Arguably not the best one, but before I continue spending more time on this I wanted to ask for a review or some feedback. I need to know if Im heading more or less in the right direction, otherwise I can stop before it costs too much time :)

- I did not add any unittests yet
- I tested the code by using a local docker setup and modifying it to work with https
- I used the docker setup for sending encrypted and signed emails from and to different local users
- I generated SMIME p12 certificates by running the following script:


````
#! /bin/sh

#umask 077

# $user is used for the subdir to hold the certs AND for
# the certificate's Common Name
user="$1"
mkdir -p "smime/$user"

#umask 277

KEY="smime/$user/cert.key"
CSR="smime/$user/cert.csr"
CERT="smime/$user/cert.crt"
P12="smime/$user/cert.p12"

# btf075@ph-freiburg.de
# certificate details for herenow script (configurable)
COUNTRY="DE"                # 2 letter country-code
STATE="Sachsen"            # state or province name
LOCALITY="Dresden"        # Locality Name (e.g. city)
ORGNAME="organisation name" # Organization Name (eg, company)
ORGUNIT=""                  # Organizational Unit Name (eg. section)
EMAIL="$user@dev.localhost"    # certificate's email address
# optional extra details
CHALLENGE="bullshit123"                # challenge password
COMPANY=""                  # company name

# generating key
openssl genrsa -des3 -out $KEY 4096


# create the csr
cat <<__EOF__ | openssl req -new -key $KEY -out $CSR
$COUNTRY
$STATE
$LOCALITY
$ORGNAME
$ORGUNIT
$user
$EMAIL
$CHALLENGE
$COMPANY
__EOF__

# Sing the key
openssl x509 -req -days 365 -in $CSR -signkey $KEY -out $CERT

# Converting to pkcs12
openssl pkcs12 -export -in $CERT -inkey $KEY -name "$user" -out $P12

````

boekhorstb1 committed:
trying to add multiple private keys

boekhorstb1 committed:
adding first version of a database

boekhorstb1 committed:
adding db adapter for smime extra keys

boekhorstb1 committed:
adding the real keys to the db tables

boekhorstb1 committed:
trying to add first frontend addition

boekhorstb1 committed:
adding extra private keys

boekhorstb1 committed:
creating first set and unset function for certificates

boekhorstb1 committed:
update with check if private key is allready set

boekhorstb1 committed:
adding deletion of extra smime keys

boekhorstb1 committed:
Adding help message

boekhorstb1 committed:
adding keys to fix import and setting of personal certificates

boekhorstb1 committed:
adding js frontend for selecting other certificate to decrypt an email

boekhorstb1 committed:
adapting decrypt function and changing privateKey method

boekhorstb1 committed:
button for loading other key is working, just need to reload the window

boekhorstb1 committed:
changing keys for decryption is working

boekhorstb1 committed:
fixing methods

boekhorstb1 committed:
basic keys switching is working now for decryption only

boekhorstb1 committed:
sw

boekhorstb1 committed:
using this instead of all the js stuff

boekhorstb1 committed:
removing the js ajax things again

boekhorstb1 committed:
adding comments

boekhorstb1 committed:
reverting js and setting simpler code to decrypt all message by using all keys available

boekhorstb1 committed:
storing passwords

boekhorstb1 committed:
adding password to db as necessary

boekhorstb1 committed:
zwischenspeicher

boekhorstb1 committed:
got it!

boekhorstb1 committed:
encrypting the password

boekhorstb1 committed:
removing addExtrPublicKey function

boekhorstb1 committed:
throwing errors instead of returning them

boekhorstb1 committed:
adding precommit and fixing tabs

boekhorstb1 committed:
adapting with is_null function

boekhorstb1 committed:
adapting docstring of getExtraPublicKey

boekhorstb1 committed:
adapting naming and adding info to docstring

boekhorstb1 committed:
removed ->message because cannot throw a string

boekhorstb1 committed:
simplifying by removing try catch and adding more docstring

boekhorstb1 committed:
removed imp5 and adapted syntax of other migrations wiht phpcs fixer

boekhorstb1 committed:
sw

boekhorstb1 committed:
removing all unnecesarry try catch blocks

boekhorstb1 committed:
change naming and formatting

boekhorstb1 committed:
sw

boekhorstb1 committed:
adding warning for deleting of certs in store

boekhorstb1 committed:
adding warning for unssets of certs

boekhorstb1 committed:
sw

boekhorstb1 committed:
adding extra import button

boekhorstb1 committed:
adding import for extra keys

boekhorstb1 committed:
adding notification and resetting check of pk

boekhorstb1 committed:
reverting accidentaly removed try catch

boekhorstb1 committed:
adding import button

boekhorstb1 committed:
sq

boekhorstb1 committed:
sqw

boekhorstb1 committed:
starting secondary keys: adding buttons and help files

boekhorstb1 committed:
fixing debug error and continuing...

boekhorstb1 committed:
solving inlinescript issues

boekhorstb1 committed:
sw

boekhorstb1 committed:
stop unsetting a key when password not entered

boekhorstb1 committed:
unsetting should not work if key is unknown or password ist not set

boekhorstb1 committed:
sw

boekhorstb1 committed:
starting with setSmimeSecondary

boekhorstb1 committed:
adding secondary key... issues remain

boekhorstb1 committed:
unsetting means delete if key in db and also adding ids of sign keys

boekhorstb1 committed:
update notifications

boekhorstb1 committed:
adding id message if not set in db

boekhorstb1 committed:
trying to set alias dialog js

boekhorstb1 committed:
save.. might need to reset this commit

boekhorstb1 committed:
fixing observer on null problem with alias code

boekhorstb1 committed:
fixing issue with null

boekhorstb1 committed:
able to set aliases now

boekhorstb1 committed:
sw

boekhorstb1 committed:
creating aliases for singkeys as well

boekhorstb1 committed:
refactoring code names

boekhorstb1 committed:
working on aliases and simplifying code

boekhorstb1 committed:
first version of setting aliases

boekhorstb1 committed:
sw

boekhorstb1 committed:
sw

boekhorstb1 committed:
adding public key view

boekhorstb1 committed:
adding info of pub certs

boekhorstb1 committed:
simpleway to add pubkeys to the prefs of an identity: remains the issue of loading keys for encryption... probably will need to allow a connection to smime.php

boekhorstb1 committed:
using SmimePrivateKey for the keypair prefs of identities

boekhorstb1 committed:
trying to add smime options to identities

boekhorstb1 committed:
sw

boekhorstb1 committed:
zwischenspeicherung

boekhorstb1 committed:
zwischenudpate 2

boekhorstb1 committed:
some template

boekhorstb1 committed:
fighting

boekhorstb1 committed:
finally got it

boekhorstb1 committed:
cleaning up code a bit

boekhorstb1 committed:
sw

boekhorstb1 committed:
updating migration with needed id values

boekhorstb1 committed:
trying to load identities in the prefs window

boekhorstb1 committed:
zwischenspeichern

boekhorstb1 committed:
simplifying by removing support for multiple keys for identities

boekhorstb1 committed:
finally understanding horde do action

boekhorstb1 committed:
zwischenspeicehr add clickable links

boekhorstb1 committed:
adding links to identity page

boekhorstb1 committed:
lists are showing

boekhorstb1 committed:
ws

boekhorstb1 committed:
set loading of keys on domload and on change of identity selected

boekhorstb1 committed:
add check and relink for standardidentity to use smime page for key management

boekhorstb1 committed:
problem: smime key should be bound to identity field in prefs table or in extra table completely

boekhorstb1 committed:
working on setting keys to identity when default id is changed

boekhorstb1 committed:
setting priv and pub keys to the identity array

boekhorstb1 committed:
getting  and setting keys to and from prefs identity array

boekhorstb1 committed:
switching between identities switches the smimekeys

boekhorstb1 committed:
adding identity options to functions so private keys of other identities are shown

boekhorstb1 committed:
fixing deletion form identity array

boekhorstb1 committed:
sw

boekhorstb1 committed:
adding  to the unset, remove and get passphrase methods

boekhorstb1 committed:
working on setting additional identities to other id than default

boekhorstb1 committed:
adding support for showing links in prefs personal inforamtion section per selected identity

boekhorstb1 committed:
taking the easy way out and removing all the js and otherstuff in favor for a link to the smime preferences page

boekhorstb1 committed:
semi fixing additional certs

boekhorstb1 committed:
possible to send f encrypt andorm difeorent identity with other identity

boekhorstb1 committed:
adding a list with identities to smime, still need to make it switchable

boekhorstb1 committed:
setting links to identity mail page

boekhorstb1 committed:
fixin decryption with multiple identities

boekhorstb1 committed:
option to choose adressbook or smime for keymanagement

boekhorstb1 committed:
fixing errors and trying to add option to view pubkeys in prefs section

boekhorstb1 committed:
trying to make js work for links in prefs

boekhorstb1 committed:
cleaning comments and code

ralflang commented:
That's quite a huge patch. I will do a step by step review later tonight but approach wise it looks good.

One ask: check all usages of $() if there is an easy replacement with Javascript builtin functionality. We will at some point drop prototypejs entirely.

ralflang committed:
Delete .pre-commit-config.yaml

This should not be part of this PR. Let's discuss separately if this seems useful.

ralflang commented on js/prefs/switchencryptionoptions.js:
```diff
@@ -0,0 +1,97 @@
+    // checks wheather the smime checkbox is clicked and shows/hides
+    checkSMIMEOption: function () {
+        // get checkbox info
+        if ($('smimeselect').checked === true) {
```
Use document.getElementById(idstring) for this type

boekhorstb1 replied:
Ok I replaced all prototypejs to normal javascript. Thanks for the email with information! :D

I think some functionality has now decreased. If I click the check box "Check if you would like to use SMIME-keypairs for identities", then the SMIME option appears but the infos on identity-keys in the addressbook remain, sometimes (!), after some more clicking they get removed..

Also: if the checkbox is unchecked, I have to wait a few seconds until code appears. Something with ajax maybe? For some reason Ajax loads 3x

ralflang commented on js/prefs/switchencryptionoptions.js:
```diff
@@ -0,0 +1,97 @@
+        let identity = $("identity");
+        let identityID = Number($F(identity));
+
+        HordeCore.doAction('getIdentityPubKey', // NOTE: methods from doAction cannot get any parameters! They expect $this->vars->$name
```
You can/should unwrap in the Ajax layer and call into a method which actually has parameters.

ralflang commented on js/prefs/switchencryptionoptions.js:
```diff
@@ -0,0 +1,97 @@
+};
+
+// on change of identity show chosen options for SMIME manament or Adressbook keys management
+document.observe('HordeIdentitySelect:change', switchEncryption.checkSMIMEOption.bindAsEventListener(switchEncryption));
```
Use document.addEventListener('eventname', function(event) { } for document.observe replacements.

boekhorstb1 replied:
ok did that thanks!

ralflang commented on lib/Ajax/Application.php:
Yes, you bury a single newly registered handler in dozens of autocorrected styles. That's not nice for review. Otherwise the file is fine.

boekhorstb1 replied:
sorry! removed pre-commit

ralflang commented on lib/Ajax/Application/Handler/Alias.php:
Please leave this file as a shell and move the actual implementation to a strongly typed class/method. Otherwise it's fine.

ralflang commented on lib/Ajax/Application/Handler/SwitchEncryption.php:
Please leave this file as a shell and move the actual implementation to a strongly typed class/method. Otherwise it's fine.

ralflang commented on lib/Ajax/Application/Handler/SwitchEncryption.php:
```diff
@@ -0,0 +1,56 @@
+        $name = $identity->getName($identityID);
+
+        $smime_url = IMP_Basic_Smime::url();
+        $self_url = Horde::url($GLOBALS['registry']->getServiceLink('prefs', 'imp'), false);
```
When factoring out, use the globals (registry, injector, notification) either as parameters passed in or use their results. We don't want globals magic inside logic.

boekhorstb1 replied:
Ok, Im trying to do that. 
So is [this](https://github.com/maintaina-com/imp/pull/9/commits/a74fe46dc035a1c1c126079a8b5da998f6175b3a) what you mean more or less? Added injector as a parameter and called all other classes by `use class...` @ralflang ?

ralflang commented on lib/Ajax/Imple/AliasDialog.php:
Imple's are ancient. I hope this works as intended. We should move all the imple's and most of the doActions to proper controller API in a separate, later PR.

boekhorstb1 replied:
Ok, I simply used existing code as an example: the set-password dialog is an Imple

ralflang commented on migration/5_imp_smime_alias.php:
Please move this into 4.

boekhorstb1 replied:
ok did that.
It might be nice to have another table for identities here. But I still have to learn how to use FKs, I think they are purely code based in Horde right? I saw some examples in some other code

ralflang reviewed (commented):
This is going in the right direction. Let's get some minor issues solved and then merge this.

boekhorstb1 committed:
converting prototype to pure js

boekhorstb1 committed:
sw

boekhorstb1 committed:
trying to leave ajax class SwitchEncryption as a shell

boekhorstb1 committed:
adding injector as parameter and calling all other classes from other than globals

boekhorstb1 committed:
fixing links

boekhorstb1 committed:
merging all newer migration files with 4_imp_smime

boekhorstb1 committed:
adding handler for alias and makeing it work with identities

boekhorstb1 committed:
fixing addition of singkey

boekhorstb1 committed:
removing specified pubkey retrival for now adding fix for managing multiple keys with multiple identies and removing unnecessary code

boekhorstb1 committed:
need migrate script to add existing smime keys from previous code

boekhorstb1 committed:
draft structure for change

boekhorstb1 committed:
zwischenspeicher

boekhorstb1 committed:
fixing key comparison

boekhorstb1 committed:
trying to update getSetkeys

boekhorstb1 committed:
getting keys from extratable update

boekhorstb1 committed:
set ids everywhere to prefs tables

boekhorstb1 committed:
issues

boekhorstb1 committed:
adapting logic to add smime keys

boekhorstb1 committed:
ws

boekhorstb1 committed:
setting prefs as set keys and using identies only as numbers

boekhorstb1 committed:
fixing stuff

boekhorstb1 committed:
fixing pub keys

boekhorstb1 committed:
smime pubkeys field removed in favor of one keyid

boekhorstb1 committed:
is this makeing the logic clearer? I hope

boekhorstb1 committed:
changing way to check keys in privatekeys

boekhorstb1 commented:
ok looks like it is working :)
When you have time, could you take a look at it @midahp @ralflang, if you have time?

I guess I would need to add unittests to this as well. Should I add those to this same MR or create a new one maybe?

boekhorstb1 committed:
changing help text and link name a little

boekhorstb1 committed:
updated translation files

boekhorstb1 committed:
fixing bugy link to identities

boekhorstb1 committed:
DE übersetzung

boekhorstb1 committed:
translation update

boekhorstb1 committed:
adding translation for js code

midahp commented on lib/Mime/Viewer/Smime.php:
```diff
@@ -155,73 +157,131 @@ protected function _getEmbeddedMimeParts()
+                }
+
+                // check if privatekeys can be retrieved
+                if (!empty($keyslist) && isset($keyslist)) {
```
I don't think you need this `if`

midahp commented on lib/Mime/Viewer/Smime.php:
```diff
@@ -155,73 +157,131 @@ protected function _getEmbeddedMimeParts()
+                            break;
+                        }
+                    }
+                } else {
```
I don't think you need this. Either it worked or it is still `null`

midahp commented on lib/Mime/Viewer/Smime.php:
```diff
@@ -155,73 +157,131 @@ protected function _getEmbeddedMimeParts()
+            /* if still nothing worked: return an error specifying why decryption did not work */
+            if (is_null($decrypted_data)) {
+                // in case all of the certificates failed to decrypt, throw an error
+                if (isset($f) && $f->getMessage() == $e->getMessage()) {
```
What is `$f` here? I could not find it

midahp commented on lib/Mime/Viewer/Smime.php:
```diff
@@ -155,73 +157,131 @@ protected function _getEmbeddedMimeParts()
      * @return mixed  See self::_getEmbeddedMimeParts().
      */
-    protected function _parseEnvelopedData()
+    protected function _parseEnvelopedData($otherkey=null)
```
I don't think you use `$otherkey` in this method

midahp commented on lib/Prefs/Identity.php:
```diff
@@ -206,6 +207,37 @@ public function getFromAddress($ident = null)
+     *
+     * @return String  A humanreadale email-address.
+     */
+    public function getEmail($ident = null)
```
`$ident` is not used

midahp commented on lib/Prefs/Special/SmimePrivateKey.php:
```diff
@@ -23,57 +23,205 @@
+     * Displays function of Horde_Core_Prefs_Ui, called after init()
      */
-    public function display(Horde_Core_Prefs_Ui $ui)
+    public function display(Horde_Core_Prefs_Ui $ui, $identityID =0)
```
`$identityID` is not used

midahp reviewed (approved):
I could not really look at it all, because it is so much, but it mostly looks good

boekhorstb1 committed:
adapting code to review of mahdi

boekhorstb1 commented:
do you think you can take the MR @ralflang ?
